File: src/main/java/io/github/kingg22/api/vacunas/panama/web/controller/UsuarioController.java
package io.github.kingg22.api.vacunas.panama.web.controller;

import io.github.kingg22.api.vacunas.panama.persistence.entity.Doctor;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Entidad;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Fabricante;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Paciente;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Permiso;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Persona;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Rol;
import io.github.kingg22.api.vacunas.panama.persistence.entity.Usuario;
import io.github.kingg22.api.vacunas.panama.response.ApiResponse;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseCode;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil;
import io.github.kingg22.api.vacunas.panama.response.DefaultApiError;
import io.github.kingg22.api.vacunas.panama.service.IUsuarioManagementService;
import io.github.kingg22.api.vacunas.panama.util.RolesEnum;
import io.github.kingg22.api.vacunas.panama.web.dto.LoginDto;
import io.github.kingg22.api.vacunas.panama.web.dto.RegisterUser;
import io.github.kingg22.api.vacunas.panama.web.dto.RestoreDto;
import jakarta.validation.Valid;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.authentication.password.CompromisedPasswordException;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.context.request.ServletWebRequest;

/**
 * Controller for {@link Usuario} registration and management, {@link Rol} and {@link Permiso}.
 *
 * <p>This controller handles operations related to registering users and managing their roles and associated entities
 * (e.g., {@link Paciente}, {@link Doctor}, {@link Fabricante}). It ensures that users are linked to an existing
 * {@link Persona} or {@link Entidad} and properly assigned roles.
 *
 * <p><b>Response Format:</b> The response for registration and related endpoints typically includes:
 *
 * <ul>
 *   <li>User details (e.g., username, roles, etc.).
 *   <li>Associated {@link Persona} or {@link Entidad} information (e.g., {@link Paciente}, {@link Doctor},
 *       {@link Fabricante}) if applicable.
 *   <li>A JWT token, which is only generated if the associated persona or entity has an active (validated) status.
 * </ul>
 *
 * <br>
 */
@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping(path = "/vacunacion/v1/account", produces = MediaType.APPLICATION_JSON_VALUE)
public class UsuarioController {
    private final AuthenticationManager authenticationManager;
    private final IUsuarioManagementService usuarioManagementService;

    /**
     * Handles user registration.
     *
     * <p>First, it checks if the current {@link Authentication} is for an authenticated user with sufficient
     * permissions to create users with lower {@link Rol}. If not authenticated, it allows registering a
     * {@link Paciente}. It also validates tha data to be registered (e.g., username, email) is not currently in use.
     *
     * <p>If all validations pass, the {@link Usuario} is created.
     *
     * <p><b>Note:</b> The user must be assigned roles, and empty roles are not allowed. If the associated entities is
     * not created, the request will be rejected. For cases where both the {@link Persona}/{@link Entidad} and the
     * {@link Usuario} need to created in a single request, a different endpoint should be used.
     *
     * @param registerUser The {@link RegisterUser} containing the user registration details.
     * @param authentication The {@link Authentication} representing the current user (if any).
     * @param request The {@link ServletWebRequest} used for building the response.
     * @return {@link ApiResponse} containing the registration result, including user details, associated
     *     {@link Persona} or {@link Entidad} information and a token if the {@link Persona} or {@link Entidad} is
     *     validated and active.
     */
    @PostMapping({"/register"})
    public ResponseEntity<ApiResponse> register(
            @RequestBody @Valid RegisterUser registerUser, Authentication authentication, ServletWebRequest request) {
        var apiResponse = ApiResponseFactory.createResponse();
        var usuarioDto = registerUser.usuario();
        if (authentication != null
                && authentication.isAuthenticated()
                && !(authentication instanceof AnonymousAuthenticationToken)) {
            var failedList = this.usuarioManagementService.validateAuthoritiesRegister(usuarioDto, authentication);
            apiResponse.addErrors(failedList);
        } else if (usuarioDto.roles() != null
                && !usuarioDto.roles().stream()
                        .allMatch(rolDto -> rolDto.id() != null
                                        && RolesEnum.getByPriority(rolDto.id()).equals(RolesEnum.PACIENTE)
                                || rolDto.nombre() != null && rolDto.nombre().equalsIgnoreCase("Paciente"))) {
            apiResponse.addError(new DefaultApiError(
                    ApiResponseCode.MISSING_ROLE_OR_PERMISSION,
                    "Solo pacientes pueden registrarse sin autenticación"));
        }

        if (apiResponse.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.FORBIDDEN);
            apiResponse.addStatus("message", ApiResponseCode.INSUFFICIENT_ROLE_PRIVILEGES);
            return ApiResponseUtil.sendResponse(apiResponse, request);
        }

        var apiContentResponse = this.usuarioManagementService.createUser(registerUser);
        apiResponse.mergeContentResponse(apiContentResponse);
        if (apiContentResponse.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.BAD_REQUEST);
        } else {
            apiResponse.addStatusCode(HttpStatus.CREATED);
        }
        return ApiResponseUtil.sendResponse(apiResponse, request);
    }

    @PostMapping({"/login"})
    public ResponseEntity<ApiResponse> login(@RequestBody @Valid LoginDto loginDto, ServletWebRequest request) {
        var apiResponse = ApiResponseFactory.createResponse();
        Authentication authentication = null;

        try {
            authentication = this.authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(loginDto.username(), loginDto.password()));
        } catch (CompromisedPasswordException exception) {
            log.debug("CompromisedPassword: {}", exception.getMessage());
            apiResponse.addStatusCode(HttpStatus.TEMPORARY_REDIRECT);
            apiResponse.addStatus("Please reset your password in the given uri", "/vacunacion/v1/account/restore");
            apiResponse.addError(new DefaultApiError(
                    ApiResponseCode.COMPROMISED_PASSWORD,
                    "password",
                    "Su contraseña está comprometida, por favor cambiarla lo más pronto posible"));
        }

        if (authentication != null && authentication.isAuthenticated()) {
            apiResponse.addData(this.usuarioManagementService.setLoginData(UUID.fromString(authentication.getName())));
            apiResponse.addStatusCode(HttpStatus.OK);
            apiResponse.addStatus("message", "Login successful");
        }

        return ApiResponseUtil.sendResponse(apiResponse, request);
    }

    @PatchMapping({"/restore"})
    public ResponseEntity<ApiResponse> restore(@RequestBody @Valid RestoreDto restoreDto, ServletWebRequest request) {
        var apiResponse = ApiResponseFactory.createResponse();
        var apiContentResponse = this.usuarioManagementService.changePassword(restoreDto);
        apiResponse.mergeContentResponse(apiContentResponse);
        if (apiContentResponse.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.BAD_REQUEST);
        } else {
            apiResponse.addStatusCode(HttpStatus.OK);
        }
        return ApiResponseUtil.sendResponse(apiResponse, request);
    }

    @GetMapping
    public ResponseEntity<ApiResponse> profile(@NotNull Authentication authentication, ServletWebRequest request) {
        var apiResponse = ApiResponseFactory.createResponse();
        apiResponse.addData(usuarioManagementService.getProfile(UUID.fromString(authentication.getName())));
        apiResponse.addStatusCode(HttpStatus.OK);
        return ApiResponseUtil.sendResponse(apiResponse, request);
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/usuario/controller/UsuarioController.kt
package io.github.kingg22.api.vacunas.panama.modules.usuario.controller

import io.github.kingg22.api.vacunas.panama.modules.usuario.dto.LoginDto
import io.github.kingg22.api.vacunas.panama.modules.usuario.dto.RegisterUserDto
import io.github.kingg22.api.vacunas.panama.modules.usuario.dto.RestoreDto
import io.github.kingg22.api.vacunas.panama.modules.usuario.service.UsuarioService
import io.github.kingg22.api.vacunas.panama.response.ActualApiResponse
import io.github.kingg22.api.vacunas.panama.response.ApiResponseCode
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory.createApiErrorBuilder
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory.createResponse
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil.createResponseEntity
import io.github.kingg22.api.vacunas.panama.util.HaveIBeenPwnedPasswordChecker
import io.github.kingg22.api.vacunas.panama.util.bcryptMatch
import io.github.kingg22.api.vacunas.panama.util.isCompromisedUsing
import io.github.kingg22.api.vacunas.panama.util.logger
import io.vertx.ext.web.RoutingContext
import jakarta.annotation.security.PermitAll
import jakarta.enterprise.context.RequestScoped
import jakarta.inject.Inject
import jakarta.validation.Valid
import jakarta.ws.rs.Consumes
import jakarta.ws.rs.GET
import jakarta.ws.rs.PATCH
import jakarta.ws.rs.POST
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import org.eclipse.microprofile.jwt.JsonWebToken
import java.util.UUID

/**
 * Controller for `Usuario` registration and management, `Rol` and `Permiso`.
 *
 * This controller handles operations related to registering users and managing their roles and associated entities
 * (e.g., `Paciente`, `Doctor`, `Fabricante`). It ensures that users are linked to an existing
 * `Persona` or `Entidad` and properly assigned roles.
 *
 * **Response Format**: The response for registration and related endpoints typically includes:
 *
 *  * User details (e.g., username, roles, etc.).
 *  * Associated `Persona` or `Entidad` information (e.g., `Paciente`, `Doctor`, `Fabricante`) if applicable.
 *  * A JWT token, which is only generated if the associated persona or entity has an active (validated) status.
 *
 * @see io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Usuario
 * @see io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Rol
 * @see io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Permiso
 * @see io.github.kingg22.api.vacunas.panama.modules.persona.entity.Persona
 * @see io.github.kingg22.api.vacunas.panama.modules.paciente.entity.Paciente
 * @see io.github.kingg22.api.vacunas.panama.modules.doctor.entity.Doctor
 * @see io.github.kingg22.api.vacunas.panama.modules.common.entity.Entidad
 * @see io.github.kingg22.api.vacunas.panama.modules.fabricante.entity.Fabricante
 */
@Path("/account")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
@RequestScoped
class UsuarioController(
    private val pwChecker: HaveIBeenPwnedPasswordChecker,
    private val usuarioService: UsuarioService,
) {
    private val log = logger()

    @Inject
    lateinit var jwt: JsonWebToken

    /**
     * Handles user registration.
     *
     * First, it checks if the current [Authentication] is for an authenticated user with sufficient
     * permissions to create users with lower `Rol`. If not authenticated, it allows registering a
     * `Paciente`. It also validates the data to be registered (e.g., username, email) is not currently in use.
     *
     * If all validations pass, the `Usuario` is created.
     *
     * **Note**: The user must be assigned roles, and empty roles are not allowed. If the associated entities are
     * not created, the request will be rejected. For cases where both the `Persona` / `Entidad` and the
     * `Usuario` need to create in a single request, a different endpoint should be used.
     *
     * @param registerUserDto The [RegisterUserDto] containing the user registration details.
     * @param authentication The [Authentication] representing the current user (if any).
     * @param request The [ServerHttpRequest] used for building the response.
     * @return [ActualApiResponse] containing the registration result, including user details, associated
     * `Persona` or `Entidad` information and a token if the `Persona` or `Entidad` is
     * validated and active.
     *
     * @see io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Usuario
     * @see io.github.kingg22.api.vacunas.panama.modules.usuario.entity.Rol
     * @see io.github.kingg22.api.vacunas.panama.modules.persona.entity.Persona
     * @see io.github.kingg22.api.vacunas.panama.modules.paciente.entity.Paciente
     * @see io.github.kingg22.api.vacunas.panama.modules.common.entity.Entidad
     */
    @Path("/register")
    @POST
    @PermitAll
    suspend fun register(@Valid registerUserDto: RegisterUserDto, rc: RoutingContext): Response {
        val apiResponse = createResponse()
        val usuarioDto = registerUserDto.usuario

        /*
                if (authentication == null &&
                    !usuarioDto.roles.all { rolDto: RolDto ->
                        rolDto.id != null &&
                            getByPriority(rolDto.id) == RolesEnum.PACIENTE ||
                            rolDto.nombre != null &&
                            rolDto.nombre.equals(RolesEnum.PACIENTE.name, ignoreCase = true)
                    }
                ) {
                    apiResponse.addError(
                        createApiErrorBuilder {
                            withCode(ApiResponseCode.MISSING_ROLE_OR_PERMISSION)
                            withMessage("Solo pacientes pueden registrarse sin autenticación")
                        },
                    )
                    apiResponse.addStatusCode(403)
                    apiResponse.addStatus("message", ApiResponseCode.INSUFFICIENT_ROLE_PRIVILEGES)
                    return createResponseEntity(apiResponse, request)
                }
         */

        apiResponse.mergeContentResponse(usuarioService.createUser(registerUserDto, null))
        if (apiResponse.hasErrors()) {
            apiResponse.addStatusCode(400)
        } else {
            apiResponse.addStatusCode(201)
        }
        return createResponseEntity(apiResponse, rc)
    }

    @Path("/login")
    @POST
    @PermitAll
    suspend fun login(@Valid loginDto: LoginDto, rc: RoutingContext): Response {
        val apiResponse = createResponse()
        val usuarioDto = usuarioService.getUsuarioByIdentifier(loginDto.username)
        if (usuarioDto == null) {
            apiResponse.addStatusCode(401)
            apiResponse.addStatus("message", "Intente nuevamente")
            return createResponseEntity(apiResponse, rc)
        }
        checkNotNull(usuarioDto.id) { "Usuario id is null after retrieve with identifier: ${loginDto.username}" }
        if (loginDto.password.isCompromisedUsing(pwChecker)) {
            log.debug("CompromisedPassword for user with identifier: {}", loginDto.username)
            apiResponse.addStatusCode(307)
            apiResponse.addStatus("Please reset your password in the given uri", "/vacunacion/v1/account/restore")
            apiResponse.addError(
                createApiErrorBuilder {
                    withCode(ApiResponseCode.COMPROMISED_PASSWORD)
                    withProperty("password")
                    withMessage("Su contraseña está comprometida, por favor cambiarla lo más pronto posible")
                },
            )
            return createResponseEntity(apiResponse, rc)
        }
        if (loginDto.password.bcryptMatch(usuarioDto.password)) {
            apiResponse.mergeContentResponse(usuarioService.getLogin(usuarioDto.id))
            apiResponse.addStatusCode(200)
            apiResponse.addStatus("message", "Login successful")
        } else {
            apiResponse.addStatusCode(403)
            apiResponse.addStatus("message", "Intente nuevamente")
        }
        return createResponseEntity(apiResponse, rc)
    }

    @Path("/restore")
    @PATCH
    @PermitAll
    suspend fun restore(@Valid restoreDto: RestoreDto, rc: RoutingContext): Response {
        val apiResponse = createResponse()
        apiResponse.mergeContentResponse(usuarioService.changePassword(restoreDto))
        if (apiResponse.hasErrors()) {
            apiResponse.addStatusCode(400)
        } else {
            apiResponse.addStatusCode(200)
        }
        return createResponseEntity(apiResponse, rc)
    }

    @GET
    suspend fun profile(rc: RoutingContext): Response {
        val apiResponse = createResponse()
        try {
            val id = checkNotNull(jwt.subject) { "Jwt subject is null" }
            apiResponse.mergeContentResponse(usuarioService.getProfile(UUID.fromString(id)))
            apiResponse.addStatusCode(200)
        } catch (e: IllegalArgumentException) {
            log.error("Error while user fetching the profile", e)
            apiResponse.addStatusCode(403)
        }
        return createResponseEntity(apiResponse, rc)
    }
}
